Some UI improvements for Infra Read Only Mode

himanshi-chhabra opened the pull request:
Signed-off-by: Himanshi Chhabra <himanshi.chhabra@msystechnologies.com>

### :nut_and_bolt: Description: What code changed, and why?
We need to show No preview image if there is no cookbook or databag or client public key available

<!-- /!\ Please ensure that you are NOT disclosing any customer information without their consent /!\ -->

### :chains: Related Resources
#4750 
### :+1: Definition of Done
Added No preview image if there is no cookbook or databag or client public key available

### :athletic_shoe: How to Build and Test the Change
```
STEP 1
inside the hab studio

[default:/src:0]# build components/automate-ui-devproxy/
[default:/src:0]# start_automate_ui_background
[default:/src:0]# start_all_services

STEP 2
open new window
go to automate UI path

$ cd components/automate-ui
and run the command 

npm run serve:hab
```
To add data https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views

 **PFA video for these changes.**

```To see No preview image make sure particular tab have no data available, for that we can make changes in code to set the length of the tab to be zero.```

Steps for tabs in which have added the No Preview:

1. Cookbook
Go To  Infrastructure -> Chef Servers -> click server name-> click on org name -> Coobook

2. Databag
Go To Infrastructure -> Chef Servers -> click server name-> click on org name -> DataBag

3. Public Key Tab
Go To Infrastructure -> Chef Servers -> click server name-> click on org name -> Clients Tab -> click on any client name -> Public Key Tab

4. Policy file
Type `feat` and there will be a feature flag -> Toggle the `Chef Infra Server Views` -> Refresh the page-> will see Policy tab 



### :white_check_mark: Checklist

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [ ] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
![Chef-Automate (4)](https://user-images.githubusercontent.com/61003053/108541415-6ff21200-7308-11eb-89a9-e804428423fa.png)
![Chef-Automate (5)](https://user-images.githubusercontent.com/61003053/108541421-71233f00-7308-11eb-81ea-a39c4d6acb6c.png)
![Chef-Automate (6)](https://user-images.githubusercontent.com/61003053/108541423-71bbd580-7308-11eb-9637-15317f573b5e.png)

https://user-images.githubusercontent.com/61003053/109790246-36d66d80-7c37-11eb-8b7b-06602f1b356f.mp4

vinay033 reviewed (approved):
LGTM, 
Thank you @himanshi-chhabra

msorens commented on components/automate-ui/src/app/modules/infra-proxy/client-details/client-details.component.html:
```diff
@@ -34,18 +34,24 @@
-            <p>Public key</p>
+          <div class="empty-section" *ngIf="!client?.client_key.public_key">
+            <img alt="No preview" src="/assets/img/no_preview.gif" />
+            <p>No public Key available</p>
```
```suggestion
            <p>No public key available</p>
```

msorens commented on components/automate-ui/src/app/components/tabs/tabs.component.scss:
```diff
@@ -53,3 +53,33 @@
+  height: 12.8px;
+  left: 937px;
+  top: 221.6px;
+  background: #000000;
```
Should use named colors from _variables.scss:
```suggestion
  background: $chef-black;
```

msorens commented on components/automate-ui/src/app/modules/infra-proxy/client-details/client-details.component.scss:
```diff
@@ -63,6 +63,32 @@ chef-table {
+  }
+
+  ::ng-deep .selected {
+    background-color: #F3F6F8;
+  }
+
+  ::ng-deep .selected::before {
+    height: 4px;
+    background-color: #3864F2;
+  }
+
+  chef-option:focus {
+    outline: none;
+  }
+
+  ::ng-deep chef-option::before {
+    height: 4px;
+    background-color: #E7E9E9;
```
Again, use named colors from _variables.scss.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/clients/clients.component.scss:
```diff
@@ -32,6 +32,7 @@ chef-loading-spinner {
   margin: 100px auto;
   width: 100px;
   z-index: 199;
+  height: 100%;
```
What is this doing? (That is, there is no way to review this change without some context.)

himanshi-chhabra replied:
@msorens yes, Actually we have added the pagination on the list for that we have added this height here.

msorens commented on components/automate-ui/src/app/components/tabs/tabs.component.scss:
```diff
@@ -53,3 +53,33 @@
+    left: 939px;
+    top: 218px;
+
+    font-family: Mulish;
```
I see lots of uses of `Muli` but no use of `Mulish` in the code base--why is a new font needed here? Or is it a typo...?

himanshi-chhabra replied:
No, it's not a typo but as per the new UX design we will use this font-family but for now, I have removed this font-family as we need to keep the font-family same as we were using earlier.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.scss:
```diff
@@ -4,6 +4,32 @@ chef-page-header {
+  }
+
+  ::ng-deep .selected {
+    background-color: #F3F6F8;
+  }
+
+  ::ng-deep .selected::before {
+    height: 4px;
+    background-color: #3864F2;
+  }
+
+  chef-option:focus {
+    outline: none;
+  }
+
+  ::ng-deep chef-option::before {
+    height: 4px;
+    background-color: #E7E9E9;
```
Same comment on hard-coded colors.

msorens commented on components/automate-ui/src/app/components/tabs/tabs.component.html:
```diff
@@ -1,8 +1,11 @@
-      {{tab.tabTitle}}
+      {{tab.tabTitle}} 
     </button>
+    <span *ngIf="tab.tabTitle === 'Policyfiles'" class="beta-tag"><span>Beta</span></span>
+    <span *ngIf="tab.tabTitle === 'Details'" class="beta-tag"><span>Beta</span></span>
+    <span *ngIf="tab.tabTitle === 'Reset Admin Key'" class="beta-tag"><span>Beta</span></span>
```
Instead of hard-coding these in the template, consider defining a constant array in the backfile code...
```
const betaTabs = ['Policyfiles', 'Details', 'Reset Admin Key'];
```
... and then in here just do this (also should not need the nested `span`):
```suggestion
    <span *ngIf="betaTabs.includes(tab.tabTitle)" class="beta-tag">Beta</span>
```

msorens commented on components/automate-ui/src/app/modules/infra-proxy/data-bags-details/data-bags-details.component.scss:
```diff
@@ -73,9 +73,28 @@ chef-loading-spinner {
     color: $chef-primary-dark;
   }
+
+  ::ng-deep .selected {
+    background-color: #F3F6F8;
+  }
+
+  ::ng-deep .selected::before {
+    height: 4px;
+    background-color: #3864F2;
+  }
+
+  chef-option:focus {
+    outline: none;
+  }
+
+  ::ng-deep chef-option::before {
+    height: 4px;
+    background-color: #E7E9E9;
+  }
```
I see this repeated over and over, which is a code smell. Consider extracting a new sub-component to encapsulate this behavior and use it here in each of these components.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-roles/infra-roles.component.html:
```diff
@@ -29,7 +29,9 @@
             </chef-td>
-            <chef-td>{{ role.description === '' ?  '...' : role.description }}</chef-td>
+            <chef-td>
+              {{ (role.description.length &gt; 30)? (role.description | slice:0:30)+'...' : role.description }}
```
It looks like you are trying to improve the UX by dealing with long values here, which I like. However, it is also changing the existing behavior in the case of the empty string. Is that intentional?

msorens commented on components/automate-ui/src/app/modules/infra-proxy/policy-files/policy-files.component.html:
```diff
@@ -2,7 +2,11 @@
-    <chef-table>
+    <div class="empty-section" *ngIf="!policyFiles.length">
+      <img alt="No preview" src="/assets/img/no_preview.gif" />
+      <p>No policyFiles available</p>
```
That capitalization seems awkward. I am not that familiar with the context but it might work better as plain English. (Do double check with @vinay033 or @vsingh-msys on this point, though.)
```suggestion
      <p>No policy files available</p>
```

msorens commented on components/automate-ui/src/app/modules/infra-proxy/data-bags-details/data-bags-details.component.html:
```diff
@@ -70,7 +70,7 @@
           </div>
           <app-page-picker
-            class="data-bag-items-list-paging"
+            class="item-list-paging"
```
Nice catch. 🎉 
However, there are a few more of the same--CSS classes being used that are never defined.
These couple are in your area, so please correct also:
https://github.com/chef/automate/blob/5f0d283159e37a6d0c98fcce4eef2a9a5c6ce2fb/components/automate-ui/src/app/modules/infra-proxy/clients/clients.component.html#L30
https://github.com/chef/automate/blob/68d5586caa32d8fa3a661b6456353f3e3f7db8b6/components/automate-ui/src/app/modules/infra-proxy/environments/environments.component.html#L39

These last ones are in different code areas, but your familiarity here might allow you to easily patch them, too; okay if not.
https://github.com/chef/automate/blob/bebe636f4016dc6b9e8451458476d9ecaa53cc1e/components/automate-ui/src/app/modules/nodes/nodes-list/nodes-list.component.html#L125
https://github.com/chef/automate/blob/f8ca770fc9c7915a9658855d5130811e5875eae1/components/automate-ui/src/app/pages/integrations/detail/integrations-detail.component.html#L74

himanshi-chhabra replied:
@msorens I have added these changes in my last commit but Idk why it's not showing outdated as I checked my files the is there for the above undefined-CSS class.

msorens reviewed (changes requested):
The key change I am requesting is extracting a sub-component to avoid a lot of repetitive code.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/cookbooks/cookbooks.component.html:
```diff
@@ -2,7 +2,11 @@
   <app-empty-state *ngIf="authFailure" moduleTitle="cookbooks" (resetKeyRedirection)="resetKeyTabRedirection($event)"></app-empty-state>
   <ng-container  *ngIf="!cookbooksListLoading && !authFailure">
-    <chef-table>
+    <div class="empty-section" *ngIf="!cookbooks.length">
+      <img alt="No preview" src="/assets/img/no_preview.gif" />
+      <p>No cookbooks available</p>
+    </div>
+    <chef-table *ngIf="cookbooks.length">
```
The html changes on this page--and all the similar pages in this PR--need two unit tests each; in this case:
 * cookbook list displays no_preview icon and hides list when list is empty
 * cookbook list displays list and hides no_preview icon when cookbook is not empty

There are many examples in the code base on how to do that; here is one:
https://github.com/chef/automate/blob/db4a921a16807fd3aeff056dce02ca266edefd35/components/automate-ui/src/app/page-components/client-runs-table/client-runs-table.component.spec.ts#L219-L222

himanshi-chhabra replied:
@msorens I am getting the below error while running the ```make unit``` command. will try to resolve that error and will update the code.
![Screenshot from 2021-02-22 18-32-11](https://user-images.githubusercontent.com/61003053/108712409-bc259800-753c-11eb-91ea-4ae00b20eb49.png)
![Screenshot from 2021-02-22 18-32-15](https://user-images.githubusercontent.com/61003053/108712415-bdef5b80-753c-11eb-92ae-a563738eacff.png)

msorens replied:
Have you tried without `sudo` ? I never need to use `sudo` when I run unit tests.

himanshi-chhabra replied:
yes, I have tried that too still getting the same error

msorens replied:
In theory, we already have the correct configuration in place for the issue you reported per [this issue](https://github.com/karma-runner/karma-chrome-launcher/issues/158#issuecomment-339265457). You will see that in karma.conf.js.
I also double-checked that the output you are getting up to the error is identical to the output I am getting on my machine, except mine runs successfully to completion.

@tarablack01 wondering if you have any ideas here?

himanshi-chhabra replied:
yes on my system it got resolved and I have added the changes

himanshi-chhabra commented:
> Nice job so far, but a few notes below.
> 
> Two key changes I am requesting:
> (1) To extract a sub-component to avoid a lot of repetitive code (see individual code note on that).
> 
> (2) Need more info in "How to Build and Test the Change".
> First, I am confused by your phrase"To update the server...". What does that mean in this context?
> But more specifically, you need to explain to the reviewer how to see a screen with data and how to see a screen without data. Right now, I have no idea how I am supposed to test this.

@msorens I have updated the PR with updated changes also updated the description to test the code.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/chef-servers-list/chef-servers-list.component.html:
```diff
@@ -33,36 +34,42 @@
+                <chef-th class="no-border-th">Name</chef-th>
+                <chef-th class="no-border-th">FQDN</chef-th>
+                <chef-th class="no-border-th">IP Address</chef-th>
+                <chef-th class="no-border-th">No Of Orgs</chef-th>
```
Not from you, I understand, but let's fix this--"No" is a poor abbreviation:
```suggestion
                <chef-th class="no-border-th">Number Of Orgs</chef-th>
```

msorens commented on components/automate-ui/src/app/components/tabs/tabs.component.ts:
```diff
@@ -11,6 +11,8 @@ export class TabsComponent {
     @Output() tabChange: EventEmitter<number> = new EventEmitter<number>();
 
     @ContentChildren(TabComponent) tabs: QueryList<TabComponent>;
+    betaTabs = ['Policyfiles', 'Details', 'Reset Admin Key'];
```
```suggestion
    const betaTabs = ['Policyfiles', 'Details', 'Reset Admin Key'];
```

msorens commented:
Regarding how to test, the part you added is the part most folks reviewing this will already know. The part you did not add is the key part. You stated, "To see No preview image make sure particular tab have no data available." But how does one make sure a tab has or does not have data?

msorens reviewed (changes requested):
Summarizing remaining work items here for convenience:
1. You added unit tests for one page but there are 3 or 4 almost identical pages that need the same two unit tests (including the change I note separately below).
2. Question on Muli vs Mulish font (see https://github.com/chef/automate/pull/4754#discussion_r579747621)
3. Make a new sub-component to encapsulate a lot of the repeated HTML/CSS (see https://github.com/chef/automate/pull/4754#discussion_r579749037)
4. Cleanup on https://github.com/chef/automate/pull/4754#discussion_r579750026 -- you had marked that resolved but had not made the changes; I am not sure what happened there, but I unresolved it.
5. Still do not know how to properly test per this comment: https://github.com/chef/automate/pull/4754#issuecomment-783529978

himanshi-chhabra commented:
> Summarizing remaining work items here for convenience:
> 
> 1. You added unit tests for one page but there are 3 or 4 almost identical pages that need the same two unit tests (including the change I note separately below).
> 2. Question on Muli vs Mulish font (see [#4754 (comment)](https://github.com/chef/automate/pull/4754#discussion_r579747621))
> 3. Make a new sub-component to encapsulate a lot of the repeated HTML/CSS (see [#4754 (comment)](https://github.com/chef/automate/pull/4754#discussion_r579749037))
> 4. Cleanup on [#4754 (comment)](https://github.com/chef/automate/pull/4754#discussion_r579750026) -- you had marked that resolved but had not made the changes; I am not sure what happened there, but I unresolved it.
> 5. Still do not know how to properly test per this comment: [#4754 (comment)](https://github.com/chef/automate/pull/4754#issuecomment-783529978)

1. Added no-preview test cases for the identical pages.
2. In Muli vs Mulish font (https://github.com/chef/automate/pull/4754#discussion_r584504682 )As i have mentioned that it's not a typo but as per the new UX design we will use this font-family but for now, I have removed this font-family as we need to keep the font-family same as we were using earlier.
3. In this (https://github.com/chef/automate/pull/4754#discussion_r579749037) added the repeated HTML/CSS to one component.
4. For this (https://github.com/chef/automate/pull/4754#discussion_r579750026)  I have added these changes in my last commit but Idk why it's not marked as outdated also I checked my files the changes are there for the above undefined-CSS class.
5. For testing, I have updated the description part step by step also I have attached a video to show a no-preview image in a particular tab.

msorens reviewed (commented):
These are still unresolved:
(3) Added the repeated HTML/CSS to one component. I see that you have deleted the repeated sections, but could you explain where it all went? It seems like it is in data-bags-details, but that is not used on the other pages, is it?  So I do not follow what you have done; help me understand.

(4) You have said you deleted "clients-list-paging", etc., but they are still present. Perhaps you have not committed those changes?

(5) You have still not explained how to test your PR. While it is useful to show a video that the code works on your machine, you must explain to the reviewer how to reproduce it for themselves. You explain how to get to each page--that is good. 😁  So now, for each such page, how does one see some data AND how does one see no data? You hint at the answer with "for that we can make changes in code to set the length of the tab to be zero." But your job is to reveal the actual changes, not just give a hint. When you look at the code, it takes you a couple seconds to know what to change because you wrote it. For me, or other reviewers, we will have to spend minutes--on each page--figuring that out; just share what you already know. 😎 

(6) A new thing I noticed: looks like there are inconsistencies in when the "no data" image appears. Take a look at these references. Some look at just the length, some also check if the thing is being loaded and some check for "searching". Now they may all be fine, but I think it is worthwhile if you could double-check all of them. Thanks!
```
clients/clients.component.html
10:    <div class="empty-section" *ngIf="!searching && !clients.length">

chef-server-details/chef-server-details.component.html
130:        <div class="empty-section" *ngIf="orgs.length === 0 && !orgsListLoading">

data-bags-list/data-bags-list.component.html
5:    <div class="empty-section" *ngIf="!dataBags.length">

data-bags-details/data-bags-details.component.html
35:          <div class="empty-section" *ngIf="!searching && !dataBagItems.length">

client-details/client-details.component.html
37:          <div class="empty-section" *ngIf="!client?.client_key.public_key">

policy-files/policy-files.component.html
5:    <div class="empty-section" *ngIf="!policyFiles.length">

chef-servers-list/chef-servers-list.component.html
68:        <div class="empty-section" *ngIf="(servers.length === 0) && !chefServersLoading">

infra-roles/infra-roles.component.html
11:    <div class="empty-section" *ngIf="!searching && !roles.length">

cookbooks/cookbooks.component.html
5:    <div class="empty-section" *ngIf="!cookbooks.length">

environments/environments.component.html
9:    <div class="empty-section" *ngIf="!searching && !environments.length">
```

himanshi-chhabra commented:
> These are still unresolved:
> (3) Added the repeated HTML/CSS to one component. I see that you have deleted the repeated sections, but could you explain where it all went? It seems like it is in data-bags-details, but that is not used on the other pages, is it? So I do not follow what you have done; help me understand.
> 
> (4) You have said you deleted "clients-list-paging", etc., but they are still present. Perhaps you have not committed those changes?
> 
> (5) You have still not explained how to test your PR. While it is useful to show a video that the code works on your machine, you must explain to the reviewer how to reproduce it for themselves. You explain how to get to each page--that is good.  So now, for each such page, how does one see some data AND how does one see no data? You hint at the answer with "for that we can make changes in code to set the length of the tab to be zero." But your job is to reveal the actual changes, not just give a hint. When you look at the code, it takes you a couple seconds to know what to change because you wrote it. For me, or other reviewers, we will have to spend minutes--on each page--figuring that out; just share what you already know. 
> 
> (6) A new thing I noticed: looks like there are inconsistencies in when the "no data" image appears. Take a look at these references. Some look at just the length, some also check if the thing is being loaded and some check for "searching". Now they may all be fine, but I think it is worthwhile if you could double-check all of them. Thanks!
> 
> ```
> clients/clients.component.html
> 10:    <div class="empty-section" *ngIf="!searching && !clients.length">
> 
> chef-server-details/chef-server-details.component.html
> 130:        <div class="empty-section" *ngIf="orgs.length === 0 && !orgsListLoading">
> 
> data-bags-list/data-bags-list.component.html
> 5:    <div class="empty-section" *ngIf="!dataBags.length">
> 
> data-bags-details/data-bags-details.component.html
> 35:          <div class="empty-section" *ngIf="!searching && !dataBagItems.length">
> 
> client-details/client-details.component.html
> 37:          <div class="empty-section" *ngIf="!client?.client_key.public_key">
> 
> policy-files/policy-files.component.html
> 5:    <div class="empty-section" *ngIf="!policyFiles.length">
> 
> chef-servers-list/chef-servers-list.component.html
> 68:        <div class="empty-section" *ngIf="(servers.length === 0) && !chefServersLoading">
> 
> infra-roles/infra-roles.component.html
> 11:    <div class="empty-section" *ngIf="!searching && !roles.length">
> 
> cookbooks/cookbooks.component.html
> 5:    <div class="empty-section" *ngIf="!cookbooks.length">
> 
> environments/environments.component.html
> 9:    <div class="empty-section" *ngIf="!searching && !environments.length">
> ```

@msorens 
3) For this I have added the duplicated CSS of cookbooks, cookbooks-details, roles, roles-details, data-bag, data-bag-details, clients, data-bag-details to its parent file ```chef-server-details.component.css```.
4) For this I haven't deleted the ```client-list-paging`` but I have added the CSS of these classes to its CSS component as I test the UI it's overlapping with the table that's why I have added margin from the top. eg: in client CSS file ```clients.component.css``` at line 52 I have added the CSS of ```client-list-paging```.
5) Will add the path from where we can update the length of a particular tab eg: `Cookbook

File Change:
1. Got to components/automate-ui/src/app-modules/infra-proxy/cookbook/cookbook.compoenent.ts 
2. ``At line 52, add new line``` **this.cookbook.length = [];**

Check UI: Go To Infrastructure -> Chef Servers -> click server name-> click on org name -> Cookbook
6).Will double-check these changes today and will update them.

himanshi-chhabra commented:
> These are still unresolved:
> (3) Added the repeated HTML/CSS to one component. I see that you have deleted the repeated sections, but could you explain where it all went? It seems like it is in data-bags-details, but that is not used on the other pages, is it? So I do not follow what you have done; help me understand.
> 
> (4) You have said you deleted "clients-list-paging", etc., but they are still present. Perhaps you have not committed those changes?
> 
> (5) You have still not explained how to test your PR. While it is useful to show a video that the code works on your machine, you must explain to the reviewer how to reproduce it for themselves. You explain how to get to each page--that is good.  So now, for each such page, how does one see some data AND how does one see no data? You hint at the answer with "for that we can make changes in code to set the length of the tab to be zero." But your job is to reveal the actual changes, not just give a hint. When you look at the code, it takes you a couple seconds to know what to change because you wrote it. For me, or other reviewers, we will have to spend minutes--on each page--figuring that out; just share what you already know. 
> 
> (6) A new thing I noticed: looks like there are inconsistencies in when the "no data" image appears. Take a look at these references. Some look at just the length, some also check if the thing is being loaded and some check for "searching". Now they may all be fine, but I think it is worthwhile if you could double-check all of them. Thanks!
> 
> ```
> clients/clients.component.html
> 10:    <div class="empty-section" *ngIf="!searching && !clients.length">
> 
> chef-server-details/chef-server-details.component.html
> 130:        <div class="empty-section" *ngIf="orgs.length === 0 && !orgsListLoading">
> 
> data-bags-list/data-bags-list.component.html
> 5:    <div class="empty-section" *ngIf="!dataBags.length">
> 
> data-bags-details/data-bags-details.component.html
> 35:          <div class="empty-section" *ngIf="!searching && !dataBagItems.length">
> 
> client-details/client-details.component.html
> 37:          <div class="empty-section" *ngIf="!client?.client_key.public_key">
> 
> policy-files/policy-files.component.html
> 5:    <div class="empty-section" *ngIf="!policyFiles.length">
> 
> chef-servers-list/chef-servers-list.component.html
> 68:        <div class="empty-section" *ngIf="(servers.length === 0) && !chefServersLoading">
> 
> infra-roles/infra-roles.component.html
> 11:    <div class="empty-section" *ngIf="!searching && !roles.length">
> 
> cookbooks/cookbooks.component.html
> 5:    <div class="empty-section" *ngIf="!cookbooks.length">
> 
> environments/environments.component.html
> 9:    <div class="empty-section" *ngIf="!searching && !environments.length">
> ```

@msorens 
6) I have checked the code for some tabs we have a search box and for some tabs, we don't have searching available. That's why it's showing different conditions for different tabs.
